### ISSUE TITLE
Installation issue: hdf5@1.12.2 in latest spack on macOS (missing MPI linker flags)

### ISSUE THREAD

**Issue by climbfuji** [LLNL/spack#37281]
### Steps to reproduce the issue

On macOS, install `hdf5@1.12.2` (probably also older versions, not tested) with latest spack code.

Note that a previous version of spack (end of 2022) does not have this problems, hdf5 compiles fine on macOS. This is due to the changes from using the MPI wrappers to the actual compilers, see screenshot.

<img width="1440" alt="Screenshot 2023-04-28 at 9 30 23 AM" src="https://user-images.githubusercontent.com/8006981/235110180-a29fc752-ebd9-49f1-9e8b-8b65d6d502cd.png">

I verified this by making the following code changes, after which hdf5 built fine:
```
diff --git a/var/spack/repos/builtin/packages/hdf5/package.py b/var/spack/repos/builtin/packages/hdf5/package.py
index ec8baf1d70..bb4ef8e852 100644
--- a/var/spack/repos/builtin/packages/hdf5/package.py
+++ b/var/spack/repos/builtin/packages/hdf5/package.py
@@ -581,13 +581,13 @@ def cmake_args(self):
         if "+mpi" in spec:
             args.extend(
                 [
-                    "-DMPI_CXX_COMPILER:PATH=%s" % spec["mpi"].mpicxx,
-                    "-DMPI_C_COMPILER:PATH=%s" % spec["mpi"].mpicc,
+                    "-DCMAKE_CXX_COMPILER:PATH=%s" % spec["mpi"].mpicxx,
+                    "-DCMAKE_C_COMPILER:PATH=%s" % spec["mpi"].mpicc,
                 ]
             )

             if "+fortran" in spec:
-                args.extend(["-DMPI_Fortran_COMPILER:PATH=%s" % spec["mpi"].mpifc])
+                args.extend(["-DCMAKE_Fortran_COMPILER:PATH=%s" % spec["mpi"].mpifc])

         # work-around for https://github.com/HDFGroup/hdf5/issues/1320
         if spec.satisfies("@1.10.8,1.13.0"):
```
Note also that this bug fix is not required for hdf5@1.14.0. But that version has many other problems that prevent us from using it.

### Error message

```
     ...

     4184    /Users/dom/scratch/spack-stack-fork_merge-20230426/cache/build_stage/spack-stage-hdf5-1.12.2-srjpu3elhzr5ffpdobj57wswhhayte4y/spack-src/src/H5private.h:1654:51: note: expanded from m
             acro 'HDvsnprintf'
     4185    #define HDvsnprintf(S, N, FMT, A) vsnprintf(S, N, FMT, A)
     4186                                                      ^~~
     4187    /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include/secure/_stdio.h:75:63: note: expanded from macro 'vsnprintf'
     4188      __builtin___vsnprintf_chk (str, len, 0, __darwin_obsz(str), format, ap)
     4189                                                                  ^~~~~~
  >> 4190    Undefined symbols for architecture x86_64:
     4191      "_MPI_Comm_c2f", referenced from:
     4192          _h5pget_fapl_mpio_c in H5Pf.c.o
     4193      "_MPI_Comm_f2c", referenced from:
     4194          _h5pset_fapl_mpio_c in H5Pf.c.o
     4195      "_MPI_Info_c2f", referenced from:
     4196          _h5pget_fapl_mpio_c in H5Pf.c.o
     4197      "_MPI_Info_f2c", referenced from:
     4198          _h5pset_fapl_mpio_c in H5Pf.c.o
     4199    ld: symbol(s) not found for architecture x86_64
  >> 4200    clang: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 4201    make[2]: *** [bin/libhdf5_f90cstub.200.1.1.dylib] Error 1
  >> 4202    make[1]: *** [fortran/src/CMakeFiles/hdf5_f90cstub-shared.dir/all] Error 2
     4203    make[1]: *** Waiting for unfinished jobs....
```

### Information on your system

```
* **Spack:** 0.20.0.dev0 (f2aa6c9337446f2fb0d60614093ebee6c0e6f2fb)
* **Python:** 3.10.8
* **Platform:** darwin-ventura-skylake
* **Concretizer:** clingo
```

### Additional information

Don't have those log files anymore after my successful install w/ the bug fix above, but can recreate if needed. It's pretty clear though that the MPI wrapper needs to be used for older hdf5 versions.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by climbfuji** [comment]
@lrknox
@brtnfld
@byrnHDF
@gheber
@hyoklee
@lkurz

FYI and thanks in advance for your time

**Comment by hyoklee** [comment]
Thank you for your report, @climbfuji ! 
I'm working on it now.

**Comment by climbfuji** [comment]
> Thank you for your report, @climbfuji !
> I'm working on it now.

Thanks very much, @hyoklee!

**Comment by hyoklee** [comment]
@climbfuji , I tried [your patch in this issue
](https://github.com/hyoklee/spack/commit/bf538294624472cf144560ebf9dd3a8ec916b61b) but I still get the same link error:

```
             kxqc65mlqpoftpc36iyq/spack-src/tools/lib/h5tools_ref.c
  >> 4085    Undefined symbols for architecture x86_64:
     4086      "_MPI_Comm_c2f", referenced from:
     4087          _h5pget_fapl_mpio_c in H5Pf.c.o
     4088      "_MPI_Comm_f2c", referenced from:
     4089          _h5pset_fapl_mpio_c in H5Pf.c.o
     4090      "_MPI_Info_c2f", referenced from:
     4091          _h5pget_fapl_mpio_c in H5Pf.c.o
     4092      "_MPI_Info_f2c", referenced from:
     4093          _h5pset_fapl_mpio_c in H5Pf.c.o
     4094    ld: symbol(s) not found for architecture x86_64
```

Here's my [GitHub workflow](https://github.com/hyoklee/spack/blob/bf538294624472cf144560ebf9dd3a8ec916b61b/.github/workflows/hdf5_mac.yml) that failed. I'm wondering how it built fine for you. Any idea?

**Comment by climbfuji** [comment]
@hyoklee Thanks for trying. Unfortunately, I cannot see the verbose build log. That one would tell us what it used as compilers for hdf5. I see that you are building openmpi@4.1.5 with spack, which I am doing as well. Let me find and add the verbose build log for hdf5 here.

**Comment by climbfuji** [comment]
This is indeed weird. Something must have gone wrong when I was building these things while traveling. I found that the successful build was for hdf5 1.14.0. Let me try again ...

**Comment by climbfuji** [comment]
Ok, here are the logs. With my patch, it works.

[hdf5_fail_success.tar.gz](https://github.com/spack/spack/files/11366033/hdf5_fail_success.tar.gz)

**Comment by hyoklee** [comment]
@climbfuji , thank you for trying again! I'll take a look at it.

**Comment by climbfuji** [comment]
@hyoklee Any updates on this one? Thanks ...

**Comment by hyoklee** [comment]
Hi, @climbfuji !

I checked your logs but could not figure out how it worked for you.
I'm attaching my GitHub Action log.
I can see some difference in CMake version and make command (gmake vs. system make).

[logs_2328.tar.gz](https://github.com/spack/spack/files/11437877/logs_2328.tar.gz)

**Comment by climbfuji** [comment]
@hyoklee These logs don't help, unfortunately, because there are no details. How about you test this manually first, run the spack install with `--verbose` so that we can see which compiler it picks up.

We are now using a modified version of my suggested bug fix, we simple set both `CMAKE_C_COMPILER` and `MPI_C_COMPILER`, same for CXX and Fortran (see lines 581-593 in https://github.com/NOAA-EMC/spack/blob/jcsda_emc_spack_stack/var/spack/repos/builtin/packages/hdf5/package.py) and it works like a charm on all platforms listed in https://spack-stack.readthedocs.io/en/latest/PreConfiguredSites.html#officially-supported-spack-stack-1-3-1-installations-subset-of-tier-1 as well as in CI on macOS w/ apple-clang and on Ubuntu w/ GNU and many other developer systems.

**Comment by hyoklee** [comment]
I turned on `-v` option for `spack install`.
Your patch still invokes Apple clang for linker on fresh machine (i.e., GitHub Action):
```
2023-05-10T16:22:02.1852340Z /Users/runner/work/spack/spack/lib/spack/env/clang/clang  -std=c99 -Wno-error=implicit-function-declaration -O3 -DNDEBUG -isysroot /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -mmacosx-version-min=12.0 -dynamiclib -Wl,-headerpad_max_install_names -current_version 201.1.0 -compatibility_version 201.0.0 -compatibility_version 200.0.0 -current_version 200.1.1 -o ../../bin/libhdf5_f90cstub.200.1.1.dylib -install_name @rpath/libhdf5_f90cstub.200.dylib "CMakeFiles/hdf5_f90cstub-shared.dir/H5f90kit.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5_f.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Af.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Df.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Ef.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Ff.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Gf.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5If.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Lf.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Of.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Pf.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Rf.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Sf.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Tf.c.o" "CMakeFiles/hdf5_f90cstub-shared.dir/H5Zf.c.o"  -Wl,-rpath,@executable_path/../lib -Wl,-rpath,@executable_path/ -Wl,-rpath,@loader_path/../lib -Wl,-rpath,@loader_path/ ../../bin/libhdf5.200.2.0.dylib 
```

Your success log indicates that your package invokes `mpicc`:
```
/Users/dom/scratch/spack-stack-fork_merge-20230426/envs/unified-env/install/apple-clang/14.0.0/openmpi-4.1.5-cnmzrp2/bin/mpicc  -std=c99 -Wno-error=implicit-function-declaration -O2 -g -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk -mmacosx-version-min=13.0 -dynamiclib -Wl,-headerpad_max_install_names -current_version 201.0.0 -compatibility_version 201.0.0 -compatibility_version 200.0.0 -current_version 200.1.0 -o ../../bin/libhdf5_hl.200.1.0.dylib -install_name @rpath/libhdf5_hl.200.dylib "CMakeFiles/hdf5_hl-shared.dir/H5DO.c.o" "CMakeFiles/hdf5_hl-shared.dir/H5DS.c.o" "CMakeFiles/hdf5_hl-shared.dir/H5IM.c.o" "CMakeFiles/hdf5_hl-shared.dir/H5LT.c.o" "CMakeFiles/hdf5_hl-shared.dir/H5LTanalyze.c.o" "CMakeFiles/hdf5_hl-shared.dir/H5LTparse.c.o" "CMakeFiles/hdf5_hl-shared.dir/H5PT.c.o" "CMakeFiles/hdf5_hl-shared.dir/H5TB.c.o" "CMakeFiles/hdf5_hl-shared.dir/H5LD.c.o"  -Wl,-rpath,@executable_path/../lib -Wl,-rpath,@executable_path/ -Wl,-rpath,@loader_path/../lib -Wl,-rpath,@loader_path/ ../../bin/libhdf5.200.2.0.dylib 
```

**Comment by climbfuji** [comment]
I don't know what to do about this. My suggested patch works without problems on all our macOS systems (we have about 20 users with different versions of macOS, xcode command line utilities = clang compilers, plus a CI test system).

**Comment by hyoklee** [comment]
https://github.com/github/roadmap/issues/620

**Comment by hyoklee** [comment]
I hoped that your success depended on macOS Ventura.
Thus, I tried macos-13 GitHub Action but failed to compile hdf5@1.12.2:

```
2023-05-22T17:52:49.7983500Z   >> 3755    Undefined symbols for architecture x86_64:
2023-05-22T17:52:49.7983780Z      3756      "_H5T__init_native", referenced from:
2023-05-22T17:52:49.7984050Z      3757          _H5T__init_package in H5T.c.o
2023-05-22T17:52:49.7984330Z      3758    ld: symbol(s) not found for architecture x86_64
2023-05-22T17:52:49.7984730Z   >> 3759    clang: error: linker command failed with exit code 1 (use -v to se
2023-05-22T17:52:49.7985010Z              e invocation)
```
I still can't figure out how it worked for you.

**Comment by hyoklee** [comment]
I'll try again when this PR is merged to HDF5 develop-1.12:

https://github.com/HDFGroup/hdf5/pull/2999

Please stay tuned.

**Comment by climbfuji** [comment]
@hyoklee Any update on this one? It's been a while, sorry forgot to follow up.

**Comment by climbfuji** [comment]
I am closing this. We are at 1.14 nowadays and don't see any of these problems on macOS with the latest versions.
